3.interpret_models/lr_plate_3_3p_cp_fs_data/feature_importances/average_feature_importances/nbconverted/calc.py

```diff
@@ -0,0 +1,314 @@
+lr = load(models_path / "lr_model.joblib")
+
+
+# ## Save Data
+
+# In[ ]:
+
+
+testdf = load(models_path / "testdf.joblib")
+le = load(models_path / "label_encoder.joblib")
+
+
+# ## Create Dataframe with coefficients for each Genotype
+
+# In[ ]:
+
+
+featdf = testdf.reset_index(drop=True)
+featdf.drop(["label", "preds"], inplace=True, axis=1)
+featdf = featdf.T
+featdf = featdf.reset_index()
+featdf = featdf.rename(columns={"index": "features"})
+featdf = pd.concat(
+    [pd.DataFrame(lr.coef_.T, columns=le.classes_.tolist()), featdf], axis="columns"
+)
+
+
+# ## Get the most important features for HET from the entire Test Set
+
+# In[ ]:
+
+
+featdf["abs_HET"] = featdf["HET"].abs()
+
+
+# ## Create dictionaries that map labels
+
+# In[ ]:
+
+
+pos_genes = ["HET", "Null", "WT"]
+label2gene = dict(zip(le.transform(pos_genes), pos_genes))
+gene2label = {gene: label for label, gene in label2gene.items()}
+
+
+# ## Create a DataFrame with all correctly predicted data
+
+# In[ ]:
+
+
+correct_filt = (
+    ((testdf["preds"] == 0) & (testdf["label"] == 0))
+    | ((testdf["preds"] == 1) & (testdf["label"] == 1))
+    | ((testdf["preds"] == 2) & (testdf["label"] == 2))
+)
+correctdf = testdf.loc[correct_filt]
+correctdf = correctdf.reset_index(drop=True)
+
+
+# ## Find the most important features by average
+
+# In[ ]:
+
+
+featimp = {}
+
+# Columns to be used as features
+kept_cols = correctdf.drop(["label", "preds"], axis="columns").columns
+
+for genotype in featdf[pos_genes]:
+    featimp[genotype] = {}  # Create a DataFrame for each genotype
+    label = gene2label[genotype]  # Get the label for the genotype
+    gene_filt = correctdf["label"] == label  # Filter to get the data for a genotype
+
+    # Product of model weights by the feature values
+    mat_imp = (
+        featdf[genotype].values
+        * correctdf.loc[gene_filt].drop(["label", "preds"], axis="columns").values
+    )
+
+    # Create the dataframe of product of model weights by the feature values
+    featimp[genotype]["featdf"] = pd.DataFrame(mat_imp, columns=kept_cols.to_list())
+
+    # Calculate quartiles and IQR:
+    q1 = featimp[genotype]["featdf"].quantile(0.25)
+    q3 = featimp[genotype]["featdf"].quantile(0.75)
+    iqr = q3 - q1
+
+    # Calculate bounds:
+    lower_bound = q1 - 1.5 * iqr
+    upper_bound = q3 + 1.5 * iqr
+
+    # Identify outliers from bounds:
+    featimp[genotype]["featdf"] = featimp[genotype]["featdf"][
+        (featimp[genotype]["featdf"] >= lower_bound)
+        & (featimp[genotype]["featdf"] <= upper_bound)
+    ]
+
+    # Select inliers
+    featimp[genotype]["featdf"] = featimp[genotype]["featdf"][
+        featimp[genotype]["featdf"] > 0
+    ]
+
+    # Calculate the mean for each (cell feature value / feature coefficient) product
+    featimp[genotype]["featnorm_avg"] = featimp[genotype]["featdf"].mean()
+
+    # Drop the means that are not a number
+    featimp[genotype]["featnorm_avg"].dropna(inplace=True)
```

Comment:
Thanks for the suggestion, I might just keep these separate for now, just because I'm trying to get in the habit of separating long chains of pandas commands.